### PR TITLE
Distributors: Merkle orchard

### PR DESCRIPTION
**Pull request by gtaschuk**
This implements the merkle "orchard" discussed on Tuesday - submitting as a proof of concept as we discuss priorities/approach (not a final implementation)

- Anyone can be a rewarder and reward anyone with any ERC-20 tokens using a merkle distribution
- Reward recipients can claim any collection of tokens in a single txn to an EOA, Balancer vault balance, or a callback contract (hopefully to invest in balancer pools) without approving the token
- Reward tokens are held in the vault
- There isn't a guarantee that the rewarder has supplied enough tokens for the distribution's merkle tree - claims will just fail if they haven't supplied sufficient tokens - we should give them a way to top this up
- As @dmf7z pointed out, a good feature to add would be allow people to claim on behalf of other people
- @TomAFrench identified a way we could save ~15k gas/claim vs the original implementation [by storing `claimed` more efficiently](https://github.com/balancer-labs/balancer-v2-monorepo/compare/merkle-orchard...merkle-orchard-bitmap-claims) :tada:

### CODE REVIEW

File: pkg/distributors/contracts/MerkleOrchard.sol [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+                "rewarder hasn't provided sufficient rewardTokens for claim"
+            );
+
+            ops[i] = IVault.UserBalanceOp({
+                asset: IAsset(address(claim.rewardToken)),
+                amount: claim.balance,
+                sender: address(this),
+                recipient: payable(recipient),
+                kind: kind
+            });
```
Comment (TomAFrench) [inline pkg/distributors/contracts/MerkleOrchard.sol:99]:
```suggestion
            ops[i] = IVault.UserBalanceOp({
                asset: IAsset(address(claim.rewardToken)),
                amount: claim.balance,
                sender: address(this),
                recipient: payable(recipient),
                kind: kind
            });
```

An issue I see with this is that for N claims we're performing N transfers. We should look to aggregate these.

Comment (gtaschuk) [reply pkg/distributors/contracts/MerkleOrchard.sol:99]: a difficulty in aggregating this is that we can't do in-memory mappings, so the aggregation logic is not particularly clean/comes at the expense of storage costs

Comment (TomAFrench) [reply pkg/distributors/contracts/MerkleOrchard.sol:99]: Had a stab at it here. Definitely isn't as pretty but I think it's worth the tradeoff: https://github.com/balancer-labs/balancer-v2-monorepo/pull/827

File: pkg/distributors/contracts/MerkleOrchard.sol [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+            require(
+                suppliedBalance[claim.rewardToken][claim.rewarder] >= claim.balance,
+                "rewarder hasn't provided sufficient rewardTokens for claim"
+            );
+
+            claimed[claim.rewardToken][claim.rewarder][claim.distribution][liquidityProvider] = true;
+
+            // Iterate through all the reward tokens we've seen so far.
+            for (uint256 j = 0; j < rewardTokens.length; i++) {
```
Comment (gtaschuk) [inline pkg/distributors/contracts/MerkleOrchard.sol:96]:
```suggestion
            for (uint256 j = 0; j < rewardTokens.length; j++) {
```

File: pkg/distributors/contracts/MerkleOrchard.sol [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+import "@balancer-labs/v2-vault/contracts/interfaces/IVault.sol";
+import "@balancer-labs/v2-vault/contracts/interfaces/IAsset.sol";
+
+import "./interfaces/IDistributor.sol";
+import "./interfaces/IDistributorCallback.sol";
+
+pragma solidity ^0.7.0;
+
+contract MerkleOrchard is IDistributor, Ownable {
```
Comment (nventuro) [inline pkg/distributors/contracts/MerkleOrchard.sol:31]: Created https://github.com/balancer-labs/balancer-v2-monorepo/issues/883

File: pkg/distributors/contracts/MerkleOrchard.sol [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    using FixedPoint for uint256;
```
Comment (nventuro) [inline pkg/distributors/contracts/MerkleOrchard.sol:32]: Created https://github.com/balancer-labs/balancer-v2-monorepo/issues/884

File: pkg/distributors/contracts/MerkleOrchard.sol [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    using SafeERC20 for IERC20;
+
+    // Recorded distributions
+    // rewardToken > rewarder > distribution > root
+    mapping(IERC20 => mapping(address => mapping(uint256 => bytes32))) public trees;
+    // rewardToken > rewarder distribution > lp > root
+    mapping(IERC20 => mapping(address => mapping(uint256 => mapping(address => bool)))) public claimed;
```
Comment (nventuro) [inline pkg/distributors/contracts/MerkleOrchard.sol:39]: Created https://github.com/balancer-labs/balancer-v2-monorepo/issues/885

File: pkg/distributors/contracts/MerkleOrchard.sol [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+import "./interfaces/IDistributor.sol";
+import "./interfaces/IDistributorCallback.sol";
+
+pragma solidity ^0.7.0;
+
+contract MerkleOrchard is IDistributor, Ownable {
+    using FixedPoint for uint256;
+    using SafeERC20 for IERC20;
+
+    // Recorded distributions
```
Comment (nventuro) [inline pkg/distributors/contracts/MerkleOrchard.sol:35]: Created https://github.com/balancer-labs/balancer-v2-monorepo/issues/886

File: pkg/distributors/contracts/MerkleOrchard.sol [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    // rewardToken > rewarder > distribution > root
+    mapping(IERC20 => mapping(address => mapping(uint256 => bytes32))) public trees;
+    // rewardToken > rewarder distribution > lp > root
+    mapping(IERC20 => mapping(address => mapping(uint256 => mapping(address => bool)))) public claimed;
+    // rewardToken > rewarder > balance
+    mapping(IERC20 => mapping(address => uint256)) public suppliedBalance;
+
+    IVault public immutable vault;
```
Comment (nventuro) [inline pkg/distributors/contracts/MerkleOrchard.sol:43]: https://github.com/balancer-labs/balancer-v2-monorepo/issues/887

File: pkg/distributors/contracts/MerkleOrchard.sol [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    function _processClaims(
+        address liquidityProvider,
+        address recipient,
+        Claim[] memory claims,
+        bool asInternalBalance
+    ) internal {
+        // We want to keep track of the number of unique tokens so we can aggregate transfers
+        uint256 numRewardTokens;
+        IERC20[] memory rewardTokens = new IERC20[](claims.length);
+        uint256[] memory rewardAmounts = new uint256[](claims.length);
```
Comment (nventuro) [inline pkg/distributors/contracts/MerkleOrchard.sol:66]: https://github.com/balancer-labs/balancer-v2-monorepo/issues/888

**Review by nventuro — APPROVED** [review]
I did a first pass and created new issues for most things I wanted to comment on. There's also a few minor extra things that will need addressing later on, but this is a good starting point.

Merging so we can continue working on individual issues in smaller, more targeted PRs.

**Pull request merged**